From 742c072fd66a5d28135554554bdb2a18ea914ed3 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Tue, 17 Sep 2019 16:37:48 +0800
Subject: [PATCH] DRVLIB_SW-1183: PRX300 serdes and xpcs driver needs update to
 improve Signal Integrity

Registered the calibrate for WAN xpcs and did the code cleanup for LAN xpcs
---
 drivers/phy/intel/phy-intel-combo.c    | 75 ++++++++--------------------------
 drivers/phy/intel/phy-intel-wan-xpcs.c | 42 +++++++++++++++++++
 2 files changed, 59 insertions(+), 58 deletions(-)

diff --git a/drivers/phy/intel/phy-intel-combo.c b/drivers/phy/intel/phy-intel-combo.c
index 0a20bc84e45d..f9ab98dcc3a7 100644
--- a/drivers/phy/intel/phy-intel-combo.c
+++ b/drivers/phy/intel/phy-intel-combo.c
@@ -53,6 +53,8 @@
 #define TX_POST_CUR_MASK		0x3F
 #define TX_POST_CUR_OFF			7
 #define PRE_OVRD_EN			13
+#define RX_ADAPT_REG			(RAWLANEN_RX_OV_IN_3 << 2)
+#define RX_ADAPT_DONE			0xC178
 
 static const char * const intel_phy_names[] = {"pcie", "xpcs", "sata"};
 
@@ -395,78 +397,35 @@ static int intel_pcie_set_clk_src(struct phy_ctx *iphy)
 static int phy_rxeq_autoadapt(struct phy *phy)
 {
 	struct phy_ctx *iphy;
-	int err;
+	int err, max_retry_cnt = 2;
 	u32 val;
 	struct device *dev;
 
 	iphy = phy_get_drvdata(phy);
 	dev = iphy->dev;
-	/* RAWLANEN_RX_OV_IN_3 */
-	/* 15		CONT_OVRD_EN = 0
-	 * 14		OFFCAN_CONT = 0
-	 * 13		ADAPT_CONT = 0
-	 * 12		ADAPT_REQ_OVRD_EN = 0
-	 * 11		ADAPT_REQ = 0
-	 * 10		REF_LD_VAL_OVRD_EN = 0
-	 * 9:4		REF_LD_VAL_OVRD = 0
-	 * 3		RX_LOS_THRSHLD_OVRD_EN = 0
-	 * 2:0		RX_LOS_THRSHLD_OVRD_VAL = 0
-	 */
-	/* ADAPT_REQ Bit 11 and ADAPT_REQ_OVRD_EN Bit 12 */
-	val = readl(iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2));
-	writel(val & ~(BIT(11) | BIT(12)), iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2));
-	/* ADAPT_REQ and ADAPT_REQ_OVRD_EN set to '11' */
-	writel(val | (BIT(11) | BIT(12)), iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2));
-
-	/* Check for RX Adaptation is done
-	 * RAWLANEN_DIG_AON_RX_ADAPT_DONE.RX_ADAPT_DONE[0]
-	 */
-	err = readl_poll_timeout(iphy->cr_base +
-		(RAWLANEN_RX_OV_IN_3 << 2), val, (!!(val & BIT(0))),
-		5, 5 * PHY_RXADAPT_POLL_CNT);
+
 	/* if auto adapt fail, run auto adapt one more time */
-	if (err) {
+	while (max_retry_cnt--) {
 		/* ADAPT_REQ Bit 11 and ADAPT_REQ_OVRD_EN Bit 12 */
-		val = readl(iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2));
-		writel(val & ~(BIT(11) | BIT(12)), iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2));
+		val = combo_phy_r32(iphy->cr_base, RX_ADAPT_REG);
+		combo_phy_w32(iphy->cr_base, val & ~(BIT(11) | BIT(12)),
+			      RX_ADAPT_REG);
 		/* ADAPT_REQ and ADAPT_REQ_OVRD_EN set to '11' */
-		writel(val | (BIT(11) | BIT(12)), iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2));
+		combo_phy_w32(iphy->cr_base, val | (BIT(11) | BIT(12)),
+			      RX_ADAPT_REG);
 
 		/* Check for RX Adaptation is done
 		 * RAWLANEN_DIG_AON_RX_ADAPT_DONE.RX_ADAPT_DONE[0]
 		 */
-		err = readl_poll_timeout(iphy->cr_base +
-			(RAWLANEN_RX_OV_IN_3 << 2), val, (!!(val & BIT(0))),
-			5, 5 * PHY_RXADAPT_POLL_CNT);
+		err = readl_poll_timeout(iphy->cr_base + RX_ADAPT_DONE, val,
+					 (!!(val & BIT(0))), 5,
+					 5 * PHY_RXADAPT_POLL_CNT);
+		if (!err)
+			break;
 	}
+
 	/* Stop RX Adaptation */
-	combo_phy_w32(iphy->cr_base, 0x0, (RAWLANEN_RX_OV_IN_3 << 2));
-	dev_dbg(dev, "RX Adaptation Turn off after done:\n");
-	dev_dbg(dev, "RAWLANEN_RX_OV_IN_3 (0x%p) = 0x%08x\n",
-		(iphy->cr_base + (RAWLANEN_RX_OV_IN_3 << 2)),
-		combo_phy_r32(iphy->cr_base, (RAWLANEN_RX_OV_IN_3 << 2)));
-
-	/* Display RX Adapt Value in PMA */
-	dev_dbg(dev, "Read back on RX Adapted Value\n");
-	/* LANEN_RX_ADPT_ATT_STAT ATT Adaptation code */
-	dev_dbg(dev, "LANEN_RX_ADPT_ATT_STAT [ (0x%p) = 0x%08x\n",
-		(iphy->cr_base + (LANEN_RX_ADPT_ATT_STAT << 2)),
-		combo_phy_r32(iphy->cr_base, (LANEN_RX_ADPT_ATT_STAT << 2)));
-
-	/* LANEN_RX_ADPT_VGA_STAT VGA Adaptation code */
-	dev_dbg(dev, "LANEN_RX_ADPT_VGA_STAT [ (0x%p) = 0x%08x\n",
-		(iphy->cr_base + (LANEN_RX_ADPT_VGA_STAT << 2)),
-		combo_phy_r32(iphy->cr_base, (LANEN_RX_ADPT_VGA_STAT << 2)));
-
-	/* LANEN_RX_ADPT_CTLE_STAT CTLE Adaptation code */
-	dev_dbg(dev, "LANEN_RX_ADPT_CTLE_STAT [ (0x%p) = 0x%08x\n",
-		(iphy->cr_base + (LANEN_RX_ADPT_CTLE_STAT << 2)),
-		combo_phy_r32(iphy->cr_base, (LANEN_RX_ADPT_CTLE_STAT << 2)));
-
-	/* LANEN_RX_ADPT_DFETAP1_STAT DFE Tap1 Adaptation code */
-	dev_dbg(dev, "LANEN_RX_ADPT_DFETAP1_STAT [ (0x%p) = 0x%08x\n",
-		(iphy->cr_base + (LANEN_RX_ADPT_DFETAP1_STAT << 2)),
-		combo_phy_r32(iphy->cr_base, (LANEN_RX_ADPT_DFETAP1_STAT << 2)));
+	combo_phy_w32(iphy->cr_base, 0x0, RX_ADAPT_REG);
 
 	if (err) {
 		dev_warn(dev, "PHY(%u:%u): RX Adaptation not done\n",
diff --git a/drivers/phy/intel/phy-intel-wan-xpcs.c b/drivers/phy/intel/phy-intel-wan-xpcs.c
index 40b99239c466..4ee9d28babff 100644
--- a/drivers/phy/intel/phy-intel-wan-xpcs.c
+++ b/drivers/phy/intel/phy-intel-wan-xpcs.c
@@ -14,6 +14,7 @@
  */
 #include <linux/kernel.h>
 #include <linux/module.h>
+#include <linux/iopoll.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
 #include <linux/phy/phy.h>
@@ -40,6 +41,10 @@
 #define TX_POST_CUR_MASK		0x3F
 #define TX_POST_CUR_OFF			7
 #define PRE_OVRD_EN			13
+#define PHY_RXADAPT_POLL_CNT	5000
+#define RAWLANEN_RX_OV_IN_3		0x3008
+#define RX_ADAPT_REG			(RAWLANEN_RX_OV_IN_3 << 2)
+#define RX_ADAPT_DONE			0xC178
 
 enum {
 	PHY_RST,
@@ -277,6 +282,42 @@ static int intel_wan_xpcs_phy_power_off(struct phy *phy)
 	return 0;
 }
 
+static int intel_wan_xpcs_phy_calibrate(struct phy *phy)
+{
+	struct intel_wan_xpcs_phy *priv = phy_get_drvdata(phy);
+	int err, max_retry_cnt = 2;
+	u32 val;
+
+	/* if auto adapt fail, run auto adapt one more time */
+	while (max_retry_cnt--) {
+		/* ADAPT_REQ Bit 11 and ADAPT_REQ_OVRD_EN Bit 12 */
+		val = intel_wan_xpcs_r32(priv->base, RX_ADAPT_REG);
+		intel_wan_xpcs_w32(priv->base, val & ~(BIT(11) | BIT(12)),
+				   RX_ADAPT_REG);
+		/* ADAPT_REQ and ADAPT_REQ_OVRD_EN set to '11' */
+		intel_wan_xpcs_w32(priv->base, val | (BIT(11) | BIT(12)),
+				   RX_ADAPT_REG);
+		/* Check for RX Adaptation is done
+		 * RAWLANEN_DIG_AON_RX_ADAPT_DONE.RX_ADAPT_DONE[0]
+		 */
+		err = readl_poll_timeout(priv->base + RX_ADAPT_DONE, val,
+					 (!!(val & BIT(0))), 5,
+					 5 * PHY_RXADAPT_POLL_CNT);
+		if (!err)
+			break;
+	}
+
+	/* Stop RX Adaptation */
+	intel_wan_xpcs_w32(priv->base, 0x0, RX_ADAPT_REG);
+
+	if (err) {
+		dev_warn(priv->dev, "RX Adaptation not done\n");
+		return -1;
+	}
+
+	return 0;
+}
+
 static int intel_wan_xpcs_phy_dt_parse(struct intel_wan_xpcs_phy *priv)
 {
 	struct device *dev = priv->dev;
@@ -331,6 +372,7 @@ static const struct phy_ops ops = {
 	.init		= intel_wan_xpcs_phy_init,
 	.power_on	= intel_wan_xpcs_phy_power_on,
 	.power_off	= intel_wan_xpcs_phy_power_off,
+	.calibrate	= intel_wan_xpcs_phy_calibrate,
 	.owner		= THIS_MODULE,
 };
 
